linkerd/app/core/src/errors/respond.rs

```diff
@@ -1,57 +1,100 @@
-use super::{ConnectTimeout, HttpError};
-use http::{header::HeaderValue, StatusCode};
-use linkerd_error::Error;
+use http::header::HeaderValue;
+use linkerd_error::{Error, Result};
 use linkerd_error_respond as respond;
-use linkerd_proxy_http::{ClientHandle, HasH2Reason};
-use linkerd_timeout::{FailFastError, ResponseTimeout};
+pub use linkerd_proxy_http::{ClientHandle, HasH2Reason};
 use pin_project::pin_project;
 use std::{
     pin::Pin,
     task::{Context, Poll},
 };
-use tonic::{self as grpc, Code};
-use tracing::{debug, warn};
+use tracing::{debug, info, info_span, warn};
 
 pub const L5D_PROXY_ERROR: &str = "l5d-proxy-error";
 
-pub fn layer() -> respond::RespondLayer<NewRespond> {
-    respond::RespondLayer::new(NewRespond(()))
+/// A strategy for responding to errors.
+pub trait HttpRescue<E> {
+    /// Attempts to synthesize a response from the given error.
+    fn rescue(&self, error: E) -> Result<SyntheticHttpResponse, E>;
+
+    /// A helper for inspecting potentially nested errors.
+    fn has_cause<C: std::error::Error + 'static>(
+        error: &(dyn std::error::Error + 'static),
+    ) -> bool {
+        // It's impractical for an error type to be so deeply nested that this recursion could
+        // overflow.
+        error.is::<C>() || error.source().map(Self::has_cause::<C>).unwrap_or(false)
+    }
 }
 
+#[derive(Clone, Debug)]
+pub struct SyntheticHttpResponse {
+    pub grpc_status: tonic::Code,
+    pub http_status: http::StatusCode,
+    pub close_connection: bool,
+    pub message: String,
```

Comment:
nit (possibly a premature optimization): I notice sometimes these strings are formatted dynamically, but sometimes they are fixed static strings that are converted to `String`s. I wonder if we want to use [`Cow<'static, str>`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) here so that we don't have to allocate a bunch of copies of the same string when it isn't dynamic?

we would then match on the `Cow` and call `HeaderValue::from_static` when the `Cow` has a static string in it (as a side note, i floated the idea to Sean of adding a `TryFrom<Cow<'static, str>>` to `http::header::HeaderValue` that would do that for us).